ltm/agent.py

```diff
@@ -1,102 +1,408 @@
         """
         session_id = str(uuid.uuid4())
-        self.session = LTMAgentSession(session_id=session_id, m_history=[])
-        if not self.convo_mem.is_empty():
-            self.convo_mem.add_separator()
+        self.session = LTMAgentSession(session_id=session_id, m_history=[], i_dict={})
+        if not self.semantic_memory.is_empty():
+            self.semantic_memory.add_separator()
         return self.session
 
+    def reply(self, user_message: str, agent_response: Optional[str] = None, cost_callback: Callable=None) -> str:
+
+        colour_print("CYAN", f"DEALING WITH USER MESSAGE: {user_message}")
+        self.now = datetime.datetime.now()
+
+        keywords = self.keywords_for_message(user_message, cost_cb=cost_callback)
+        context = self.create_context(user_message, max_prompt_size=self.max_prompt_size, previous_interactions=0, cost_cb=cost_callback)
+        response_text = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_callback, temperature=self.temperature)
+        debug_actions(context, self.temperature, response_text, self.llm_call_idx, self.debug_level, self.save_name,  name_template="reply-{idx}")
+        self.llm_call_idx += 1
+
+        # Save interaction to memory
+        um = Message(role="user", content=user_message, timestamp=self.now.timestamp())
+        am = Message(role="assistant", content=response_text, timestamp=self.now.timestamp())
+
+        self.save_interaction(user_message, response_text, keywords)
+        self.session.add_interaction((um, am))
+
+        return response_text
+
+    def keywords_for_message(self, user_message, cost_cb):
+
+        prompt = 'Create two keywords to describe the topic of this message:\n"{user_message}".\n\nFocus on the topic and tone of the message. Produce the keywords in JSON like: `["keyword_1", "keyword_2"]`\n\nChoose keywords that would aid in retriving this message from memory in the future.\n\nReuse these keywords if appropriate: {keywords}'
+
+        context = [make_system_message(prompt.format(user_message=user_message, keywords=self.defined_kws))]
+        while True:
+            try:
+                print("Keyword gen")
+                response = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_cb, temperature=self.temperature)
+
+                keywords = [k.lower() for k in sanitize_and_parse_json(response)]
+                break
+            except Exception as e:
+                print(repr(e) + response)
+                continue
+
+        # Update known list of keywords
+        for k in keywords:
+            if k not in self.defined_kws:
+                self.defined_kws.append(k)
+
+        print(f"Interaction keywords: {keywords}")
+        return keywords
+
+    def create_context(self, user_message, max_prompt_size, previous_interactions, cost_cb):
+
+        stamped_user_message = str(self.now) + ": " + user_message
```

Comment:
This timestamp representation is different from the one in line 193, and also from the one used in the previous baseline:
```text
[2024-07-31 12:00:56 (7 minutes and 4 seconds ago)]
<user_message_here>
```